flow/scripts/synth.tcl

```diff
@@ -63,6 +63,55 @@ if {[info exist ::env(BLOCKS)]} {
   }
 }
 
+set constr [open $::env(OBJECTS_DIR)/abc.constr w]
+puts $constr "set_driving_cell $::env(ABC_DRIVER_CELL)"
+puts $constr "set_load $::env(ABC_LOAD_IN_FF)"
+close $constr
+
+if { [info exist ::env(SYNTH_HIER_AREA_RECOVER)]  && $::env(SYNTH_HIER_AREA_RECOVER) == 1 } {
+  # Hierarchical synthesis first
+  synth  -top $::env(DESIGN_NAME)
+  opt -purge
+  #extract_fa
+  # map full adders
+  #techmap -map $::env(ADDER_MAP_FILE)
+  techmap
+  if {[info exist ::env(DFF_LIB_FILE)]} {
+    dfflibmap -liberty $::env(DFF_LIB_FILE)
+  } else {
+    dfflibmap -liberty $::env(DONT_USE_SC_LIB)
+  }
+  abc -liberty $::env(DONT_USE_SC_LIB) \
+      -constr $::env(OBJECTS_DIR)/abc.constr
+  opt_clean -purge
+
+  # Create argument list for stat
+  set stat_libs ""
+  foreach lib $::env(DONT_USE_LIBS) {
+    append stat_libs "-liberty $lib "
+  }
+  tee -o $::env(REPORTS_DIR)/synth_hier_stat.txt stat {*}$stat_libs
+
+  set ungroupThreshold 0
+  if { [info exists $::env(MAX_UNGROUP_SIZE)] } {
+    set ungroupThreshold $::env(MAX_UNGROUP_SIZE)
+  }
+  set fptr [open $::env(REPORTS_DIR)/synth_hier_stat.txt r]
+  set contents [read -nonewline $fptr]
+  close $fptr
+  set splitCont [split $contents "\n"]
+  foreach line $splitCont {
+    if {[regexp { +Chip area for module '(\S+)': (.*)} $line -> moduleName area]} {
+        if {[expr $area > $ungroupThreshold]} {
+           select -module $moduleName
+           setattr -mod -set keep_hierarchy 1
+           select -clear
+           puts "Preserving hierarchical module: $moduleName"
+        }
+    }
+  }
+}
+
```

Comment:
This just seems to double the synthesis time. What is the purpose/intent of this? If it's just to get an estimate of the module sizes, you could always run yosys normally without the flatten option (that way we only have to maintain a single core synthesis flow), and then call yosys again setting the `PRESERVE_CELLS` option to the cells exceeding your threshold.

Reply:
All these changes are to enable hierarchical synthesis instead of flat synthesis. This would allow for larger designs processing and also in line with standard synthesis practice. Now, these changes are for automated ungroup support in hierarchical synthesis flow. User would set ungroup size in Makefile or design specific config file. Then first synthesis is to get the size of modules. The modules which exceed size threshold, would be preserved and rest flattened. By default, if hierarchical synthesis is enabled (using SYNTH_HIER_AREA_RECOVER), all modules would be preserved. I was avoiding to mix with PRESERVE_CELL setting since that is user controlled and this is auto-internally generated list.

Reply:
I don't think this ends up doing what you want it to do. Once the design is mapped with abc, yosys cannot continue to optimize (it just sees blackboxes everywhere). So only the first synthesis will do something, everything else just falls through more or less.
I think you would need to either find a way to merge with the rest of the flow and run abc twice, or after this has run you need to remove it from memory and replace with new read in of the verilog.

Reply:
I've now put hierarchy generation in separate yosys run.